clippy_lints/src/unneeded_pub_crate.rs

```diff
@@ -0,0 +1,381 @@
+    }
+
+    fn visit_variant_data(
+        &mut self,
+        vd: &'tcx hir::VariantData<'tcx>,
+        _name: Symbol,
+        _generics: &'tcx hir::Generics<'tcx>,
+        parent_id: hir::HirId,
+        _span: Span,
+    ) {
+        if let Some(hir_id) = vd.ctor_hir_id() {
+            self.examine_use(self.tcx.hir().local_def_id(hir_id).to_def_id(), parent_id)
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx hir::Expr<'tcx>) {
+        match expr.kind {
+            hir::ExprKind::MethodCall(..) => {
+                self.maybe_typeck_results
+                    .and_then(|typeck_results| typeck_results.type_dependent_def(expr.hir_id))
+                    .map(|(_kind, def_id)| self.examine_use(def_id, expr.hir_id));
+            },
+            hir::ExprKind::Struct(_qpath, fields, _base) => {
+                if let Some(ty::Adt(adt_def, _substs)) = self
+                    .maybe_typeck_results
+                    .map(|typeck_results| typeck_results.expr_ty(expr).kind())
+                {
+                    self.examine_use(adt_def.did, expr.hir_id);
+                    let ty_fields = adt_def
+                        .all_fields()
+                        .map(|f| (f.ident, f))
+                        .collect::<FxHashMap<Ident, _>>();
+                    for field in fields.iter() {
+                        if let Some(ty_field) = ty_fields.get(&field.ident) {
+                            self.examine_use(ty_field.did, expr.hir_id);
+                        }
+                    }
+                }
+            },
+            hir::ExprKind::Field(base, field) => {
+                if let Some(ty::Adt(adt_def, _substs)) = self
+                    .maybe_typeck_results
+                    .map(|typeck_results| typeck_results.expr_ty(base).kind())
+                {
+                    self.examine_use(adt_def.did, expr.hir_id);
+                    if let Some(our_field) = adt_def.all_fields().filter(|f| f.ident == field).next() {
+                        self.examine_use(our_field.did, expr.hir_id);
+                    }
+                }
+            },
+            _ => {},
+        };
+        intravisit::walk_expr(self, expr);
+    }
+
+    fn visit_trait_ref(&mut self, trait_ref: &'tcx hir::TraitRef<'tcx>) {
+        if let Some(trait_ref) = trait_ref.trait_def_id() {
+            self.examine_use(trait_ref, CRATE_HIR_ID);
+        }
+        intravisit::walk_trait_ref(self, trait_ref);
+    }
+
+    fn visit_qpath(&mut self, qpath: &'tcx hir::QPath<'tcx>, hir_id: hir::HirId, span: Span) {
+        let def = match qpath {
+            hir::QPath::Resolved(_, path) => match path.res {
+                Res::Def(kind, def_id) => Some((kind, def_id)),
+                _ => None,
+            },
+            hir::QPath::TypeRelative(..) | hir::QPath::LangItem(..) => self
+                .maybe_typeck_results
+                .and_then(|typeck_results| typeck_results.type_dependent_def(hir_id)),
```

Comment:
This is the line where the ICE appears to happen. I'm not deep enough into `typeck_results` to know where the problem stems from, but I think I've seen something similar in the past. I'll need a bit of time to check, though.

Edit: It's likely because you call `typeck_results` from a visitor – usually you want to do this directly in the lint pass (which as you know is a visitor of sorts by itself).

Reply:
@llogiq i run this visitor in the context of the lint pass, down in `check_crate_post`. is that what you mean?

Reply:
my best guess is that there are qpaths that i'm not supposed to look up...

Reply:
I'm not completely sure, but I'd hope rewriting this into a single pass (where you'd collect refs to items and references to them, and finally linting all non-publicly ref'd ones) might fix it.